File: funcscript/funcs/list/SkipFunction.cs
using funcscript.core;
using funcscript.model;

namespace funcscript.funcs.list
{
    public class SkipFunction : IFsFunction
    {
        public int MaxParsCount => 2;

        public CallType CallType => CallType.Prefix;

        public string Symbol => "Skip";

        public int Precidence => 0;

        class ParList : IParameterList
        {
            public object X;
            public object N;

            public int Count => 2;

            public object GetParameter(IFsDataProvider provider, int index)
            {
                return index switch
                {
                    0 => X,
                    1 => N,
                    _ => null,
                };
            }
        }

        public object Evaluate(IFsDataProvider parent, IParameterList pars)
        {
            if (pars.Count != this.MaxParsCount)
                throw new error.EvaluationTimeException(
                    $"{this.Symbol} function: Invalid parameter count. Expected {this.MaxParsCount}, but got {pars.Count}");

            var par0 = pars.GetParameter(parent, 0);
            var par1 = pars.GetParameter(parent, 1);

            if (par0 == null)
                return null;

            if (!(par0 is FsList))
                throw new error.TypeMismatchError(
                    $"{this.Symbol} function: The first parameter should be {this.ParName(0)}");

            if (!(par1 is int))
                throw new error.TypeMismatchError(
                    $"{this.Symbol} function: The second parameter should be {this.ParName(1)}");

            var lst = (FsList)par0;
            int n = (int)par1;

            if (n <= 0)
                return lst;

            if (n >= lst.Data.Length)
                return new FsList(new object[] { });

            int newArrayLength = lst.Data.Length - n;
            object[] newArray = new object[newArrayLength];

            Array.Copy(lst.Data, n, newArray, 0, newArrayLength);

            return new FsList(newArray);
        }

        public string ParName(int index)
        {
            switch (index)
            {
                case 0:
                    return "List";
                case 1:
                    return "Number";
                default:
                    return "";
            }
        }
    }
}

File: js-port/fs-js/src/funcs/list/skip-function.js
const { BaseFunction, CallType } = require('../../core/function-base');
const { ArrayFsList } = require('../../model/fs-list');
const helpers = require('../helpers');
const { FSDataType } = require('../../core/fstypes');

class SkipFunction extends BaseFunction {
  constructor() {
    super();
    this.symbol = 'Skip';
    this.callType = CallType.Prefix;
  }

  get maxParameters() {
    return 2;
  }

  evaluate(provider, parameters) {
    const error = helpers.expectParamCount(this.symbol, parameters, this.maxParameters);
    if (error) {
      return error;
    }

    const list = helpers.ensureList(parameters.getParameter(provider, 0));
    if (!list) {
      return helpers.makeError(
        helpers.FsError.ERROR_TYPE_MISMATCH,
        `${this.symbol} function: The first parameter should be ${this.parName(0)}`
      );
    }

    const countResult = helpers.requireInteger(this.symbol, parameters.getParameter(provider, 1), 'count');
    if (!countResult.ok) {
      return countResult.error;
    }

    const count = Math.max(0, countResult.value);
    const values = [];
    for (let i = count; i < list.length; i += 1) {
      values.push(list.get(i));
    }
    return helpers.makeValue(FSDataType.List, new ArrayFsList(values));
  }

  parName(index) {
    return index === 0 ? 'List' : index === 1 ? 'Number' : '';
  }
}

module.exports = {
  SkipFunction
};
